In [22]:
data = read.table('/content/adult.csv', sep = ',', header = TRUE)

head(data)

age workclass fnlwgt education    educational.num marital.status    
1 25  Private   226802 11th          7              Never-married     
2 38  Private    89814 HS-grad       9              Married-civ-spouse
3 28  Local-gov 336951 Assoc-acdm   12              Married-civ-spouse
4 44  Private   160323 Some-college 10              Married-civ-spouse
5 18  ?         103497 Some-college 10              Never-married     
6 34  Private   198693 10th          6              Never-married     
  occupation        relationship  race  gender capital.gain capital.loss
1 Machine-op-inspct Own-child     Black Male      0         0           
2 Farming-fishing   Husband       White Male      0         0           
3 Protective-serv   Husband       White Male      0         0           
4 Machine-op-inspct Husband       Black Male   7688         0           
5 ?                 Own-child     White Female    0         0           
6 Other-service     Not-in-family White Male      0         0           
  hours.per.week native.country income
1 40             United-States  <=50K 
2 50             United-States  <=50K 
3 40             United-States  >50K  
4 40             United-States  >50K  
5 30             United-States  <=50K 
6 30             United-States  <=50K

Listing of attributes:

>50K, <=50K.

age: continuous.
workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
fnlwgt: continuous.
education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
education-num: continuous.
marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
sex: Female, Male.
capital-gain: continuous.
capital-loss: continuous.
hours-per-week: continuous.
native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

## Zad 1
### W kolumnie type_employer:
#### 1. zastąpić wpisy "Federal-gov" i "Local-gov" wpisem "SL-gov"
#### 2. zastąpić wpisy "Self-emp-inc" i "Self-emp-not-inc" wpisem "self-emp"

In [7]:
install.packages('caret')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘listenv’, ‘parallelly’, ‘future’, ‘globals’, ‘shape’, ‘future.apply’, ‘numDeriv’, ‘progressr’, ‘SQUAREM’, ‘diagram’, ‘lava’, ‘prodlim’, ‘proxy’, ‘iterators’, ‘Rcpp’, ‘clock’, ‘gower’, ‘hardhat’, ‘ipred’, ‘timeDate’, ‘e1071’, ‘foreach’, ‘ModelMetrics’, ‘plyr’, ‘pROC’, ‘recipes’, ‘reshape2’




In [5]:
library(tidyverse)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.4.2     ✔ purrr   1.0.1
✔ tibble  3.2.1     ✔ dplyr   1.1.2
✔ tidyr   1.3.0     ✔ stringr 1.5.0
✔ readr   2.1.4     ✔ forcats 1.0.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



In [8]:
library(caret)

Loading required package: lattice


Attaching package: ‘caret’


The following object is masked from ‘package:purrr’:

    lift




In [23]:
table(data$workclass)


               ?      Federal-gov        Local-gov     Never-worked 
            2799             1432             3136               10 
         Private     Self-emp-inc Self-emp-not-inc        State-gov 
           33906             1695             3862             1981 
     Without-pay 
              21 

In [24]:
data$workclass = str_replace(data$workclass, c("Local-gov|Federal-gov"), "SL-gov")
data$workclass = str_replace(data$workclass, c("Self-emp-inc|Self-emp-not-inc"), "self-emp")
table(data$workclass)


           ? Never-worked      Private     self-emp       SL-gov    State-gov 
        2799           10        33906         5557         4568         1981 
 Without-pay 
          21 

## Zad 2
### W kolumnie "marital" zredukować liczbę wpisów do trzech (Married;Not-Married;Never-Married)

In [25]:
table(data$marital.status)


             Divorced     Married-AF-spouse    Married-civ-spouse 
                 6633                    37                 22379 
Married-spouse-absent         Never-married             Separated 
                  628                 16117                  1530 
              Widowed 
                 1518 

In [26]:
data$marital.status = str_replace(data$marital.status, c("Married-AF-spouse|Married-civ-spouse|Married-spouse-absent|Separated"), "Married")
data$marital.status = str_replace(data$marital.status, c("Divorced|Widowed"), "Not-Married")
table(data$marital.status)


      Married Never-married   Not-Married 
        24574         16117          8151 

## Zad 3
### Zmniejszyć liczbę wpisów w kolumnie country (np. grupowanie przez kontynenty? inne podejście?)

In [27]:
data %>%
    group_by(native.country) %>%
    summarise(disp = mean(age))

native.country             disp    
1  ?                          38.77246
2  Cambodia                   36.89286
3  Canada                     44.04945
4  China                      41.85246
5  Columbia                   39.45882
6  Cuba                       46.35507
7  Dominican-Republic         37.97087
8  Ecuador                    37.66667
9  El-Salvador                33.38065
10 England                    40.52756
11 France                     40.31579
12 Germany                    38.60194
13 Greece                     45.83673
14 Guatemala                  32.09091
15 Haiti                      38.60000
16 Holand-Netherlands         32.00000
17 Honduras                   35.05000
18 Hong                       34.23333
19 Hungary                    50.36842
20 India                      38.36424
21 Iran                       38.37288
22 Ireland                    38.48649
23 Italy                      45.41905
24 Jamaica                    37.14151
25 Japan                      37.35870
26 Laos                       35.21739
27 Mexico                     33.63512
28 Nicaragua                  36.28571
29 Outlying-US(Guam-USVI-etc) 38.82609
30 Peru                       36.43478
31 Philippines                39.63390
32 Poland                     42.75862
33 Portugal                   41.23881
34 Puerto-Rico                39.86413
35 Scotland                   46.76190
36 South                      38.09565
37 Taiwan                     34.18462
38 Thailand                   37.66667
39 Trinadad&Tobago            39.25926
40 United-States              38.69869
41 Vietnam                    34.61628
42 Yugoslavia                 40.47826

## Zad 4
### Zastąpić wpisy "?" na wartości NA

In [28]:
table(data$workclass)


           ? Never-worked      Private     self-emp       SL-gov    State-gov 
        2799           10        33906         5557         4568         1981 
 Without-pay 
          21 

In [29]:
data[data == '?'] = NA
table(data$workclass)


Never-worked      Private     self-emp       SL-gov    State-gov  Without-pay 
          10        33906         5557         4568         1981           21 

## Zad 5
### Usunąć wiersze zawierające wpisy NA

In [30]:
data = drop_na(data)

In [31]:
table(data$workclass)


    Private    self-emp      SL-gov   State-gov Without-pay 
      33307        5442        4506        1946          21 

In [32]:
head(data)

age workclass fnlwgt education    educational.num marital.status
1 25  Private   226802 11th          7              Never-married 
2 38  Private    89814 HS-grad       9              Married       
3 28  SL-gov    336951 Assoc-acdm   12              Married       
4 44  Private   160323 Some-college 10              Married       
5 34  Private   198693 10th          6              Never-married 
6 63  self-emp  104626 Prof-school  15              Married       
  occupation        relationship  race  gender capital.gain capital.loss
1 Machine-op-inspct Own-child     Black Male      0         0           
2 Farming-fishing   Husband       White Male      0         0           
3 Protective-serv   Husband       White Male      0         0           
4 Machine-op-inspct Husband       Black Male   7688         0           
5 Other-service     Not-in-family White Male      0         0           
6 Prof-specialty    Husband       White Male   3103         0           
  hours.per.week native.country income
1 40             United-States  <=50K 
2 50             United-States  <=50K 
3 40             United-States  >50K  
4 40             United-States  >50K  
5 30             United-States  <=50K 
6 32             United-States  >50K

In [33]:
data$income = as.character(data$income)
data$income[data$income == "<=50K"] = 0
data$income[data$income == ">50K"] = 1
data['income'] = as.factor(data$income)

In [34]:
data$age = as.numeric(data$age)
data$capital.gain = as.numeric(data$capital.gain)
data$fnlwgt = as.numeric(data$fnlwgt)
data$capital.loss = as.numeric(data$capital.loss)
data$hours.per.week = as.numeric(data$hours.per.week)
data$educational.num = as.numeric(data$educational.num)

In [35]:
data$gender = as.factor(data$gender)
data$race = as.factor(data$race)
data$education = as.factor(data$education)
data$marital.status = as.factor(data$marital.status)
data$workclass = as.factor(data$workclass)

## Zad 6
### Podzielić dane na zestaw testowy i uczący

In [36]:
y = data$income

In [37]:
data_index = createDataPartition(
  y,
  times = 1,
  p = 0.7,
  list = FALSE
)

In [38]:
train_set = data[data_index, ]
test_set = data[-data_index, ]

## Zad 7
### Zbudować model za pomocą funkcji (glm)

In [39]:
formula = income ~ .

In [40]:
model = glm(formula, data = train_set, family = "binomial")

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


In [41]:
summary(model)


Call:
glm(formula = formula, family = "binomial", data = train_set)

Coefficients: (1 not defined because of singularities)
                                           Estimate Std. Error z value Pr(>|z|)
(Intercept)                              -5.327e+00  6.585e-01  -8.089 6.01e-16
age                                       2.499e-02  1.648e-03  15.168  < 2e-16
workclassself-emp                        -3.038e-01  5.330e-02  -5.699 1.20e-08
workclassSL-gov                           8.265e-02  5.949e-02   1.389 0.164723
workclassState-gov                       -2.972e-01  8.924e-02  -3.330 0.000867
workclassWithout-pay                     -4.273e-01  8.134e-01  -0.525 0.599363
fnlwgt                                    7.832e-07  1.699e-07   4.611 4.01e-06
education11th                             2.522e-03  2.083e-01   0.012 0.990341
education12th                             3.269e-01  2.714e-01   1.205 0.228249
education1st-4th                         -1.161e+00  5.987e-01  -1.940 0.05

## Zad 8
### Dopracować model za pomocą funkcji step

In [42]:
step_model = step(model)

Start:  AIC=20705.59
income ~ age + workclass + fnlwgt + education + educational.num + 
    marital.status + occupation + relationship + race + gender + 
    capital.gain + capital.loss + hours.per.week + native.country



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=20705.59
income ~ age + workclass + fnlwgt + education + marital.status + 
    occupation + relationship + race + gender + capital.gain + 
    capital.loss + hours.per.week + native.country



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”
Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


                 Df Deviance   AIC
<none>                 20526 20706
- race            4    20551 20723
- fnlwgt          1    20547 20725
- native.country 40    20638 20738
- marital.status  2    20568 20744
- workclass       4    20574 20746
- gender          1    20636 20814
- age             1    20758 20936
- capital.loss    1    20845 21023
- hours.per.week  1    20890 21068
- occupation     13    21161 21315
- relationship    5    21198 21368
- education      15    21531 21681
- capital.gain    1    22363 22541


## Zad 9
### Przedataować model na danych dataowych (tablica pomyłek)

In [43]:
predictions = predict(step_model, newdata = test_set)
y_hat = ifelse(predictions > 0.5, ">50K", "<=50K")
pred = table(predicted = y_hat, actual = test_set$income)
pred

         actual
predicted    0    1
    <=50K 9762 1736
    >50K   442 1626

In [44]:
install.packages("pROC")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [45]:
library(pROC)
roc_obj = roc(test_set$income, predictions)
roc_obj

Type 'citation("pROC")' for a citation.


Attaching package: ‘pROC’


The following objects are masked from ‘package:stats’:

    cov, smooth, var


Setting levels: control = 0, case = 1

Setting direction: controls < cases




Call:
roc.default(response = test_set$income, predictor = predictions)

Data: predictions in 10204 controls (test_set$income 0) < 3362 cases (test_set$income 1).
Area under the curve: 0.901

## Zad 10
### Obliczyć F1-score dla opracowanego modelu

In [46]:
TP <- pred[1,1]
FP <- pred[1,2]
FN <- pred[2,1]
TN <- pred[2,2]

TPR <- TP/(TP+FN)
TNR <- TN/(TN+FP)
PPV <- TP/(TP+FP)

(F_wynik = 1/(0.5*(1/TPR + 1/PPV)))
# 0.899640586121095


[1] 0.8996406

## Wyniki i podsumowanie

### pROC
Data: predictions in 10204 controls (test_set$income 0) < 3362 cases (test_set$income 1).

Area under the curve: 0.901


Wysoka wartość obszaru pod krzywą (0,901) wskazuje na wysoką jakość klasyfikatora. 

Krzywa ROC pokazuje zależność między czułością (recall) i swoistością (1 - procent wyników fałszywie dodatnich) w zależności od różnych progów używanych do klasyfikacji.

### Wynik F
Wynik F jest wysoki (0,899640586121095) - pokazuje, że wydajność modelu jest bardzo dobra i możemy przewidzieć 89% zmienności.

Wyższa wartość F wskazuje na lepszą wydajność modelu w rozpoznawaniu zarówno pozytywnych, jak i negatywnych przypadków. Wartość F może wynosić od 0 do 1, gdzie 1 oznacza doskonałą wydajność, a 0 oznacza najgorszą wydajność.

### Podsumowanie
Funkacja Step pomogła mi wybrać zmienne, które mają najlepszy wpływ na model, a pozostawienie tylko tych pomogło dostroić model tak, aby miał świetną wydajność.